### PR TITLE
fix(executor): add executable permission to staged `script`

### LINKED ISSUE #3496: Error: `exec: "/argo/staging/script": permission denied`
# Summary

I want to stop this error:

```
failed to create containerd task: OCI runtime create failed: container_linux.go:349: starting container process caused "exec: \"/argo/staging/script\": permission denied": unknown
```

# Motivation

It is annoying to need to specify `command: [sh]` for scripts, when it could just be Unix executable.

```yaml
      script:
        image: 'docker/whalesay:latest'
        command: [sh]
        source: |
          true
```

```yaml
      script:
        image: 'docker/whalesay:latest'
        source: |
          #!/bin/sh
          true
```

# Proposal

Using chmod.

---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by vatine**
Fixes #3496 

### Motivation

As-was, staged scripts end up with the permission `0644` `(rw-r--r--)` which is less than useful for shell scripts. With this modification in place, the scripts are executable.

### Modifications

Changed file permissions for the staged script.

### Verification

Built a docker image of `argoexec`, changed the workflow-controller flags to use the private image and verified that pod logs no longer had a `/argo/staging/script: permission denied` in them

### CODE REVIEW

**Comment by agilgur5** [comment]
> I must say that I am not entirely sure how the test failures relate to this change. They seem to relate to completely different things.

Hmm seems to be the same one just mentioned [on Slack](https://cloud-native.slack.com/archives/C0510EUH90V/p1710260573672209) in #12713. Indeed seems unrelated

**Review by agilgur5 — COMMENTED** [review]
> verified that pod logs no longer had a `/argo/staging/script: permission denied` in them

interestingly I've not seen this error before -- I wonder why it just popped up for you and hasn't popped up before?

**Comment by vatine** [comment]
Even more interestingly, I get the permission error for some images, but not for others (for the ones where this change was required, I also needed to include "shebang" lines listing an interpreter). I have not managed to nail down what the actual difference between these are.

I can have a look at making the permission/mode value variable and setting it to 0o755 only for scripts.

File: workflow/executor/executor.go [hunk 1]
```diff
@@ -283,7 +283,7 @@ func (we *WorkflowExecutor) StageFiles() error {
 	default:
 		return nil
 	}
-	err := os.WriteFile(filePath, body, 0o644)
+	err := os.WriteFile(filePath, body, 0o755)
```
[the remark this answers is not in the document]
Comment (vatine) [reply workflow/executor/executor.go]: I have just made a change, making this conditioned on "is a resource or script".

**Comment by agilgur5** [comment]
> Even more interestingly, I get the permission error for some images, but not for others (for the ones where this change was required, I also needed to include "shebang" lines listing an interpreter)

This lines up with what I'm suspecting, which is that this might be a [shell form vs. exec form](https://docs.docker.com/reference/dockerfile/#shell-and-exec-form) instance or something similar. Like if you have a script without executable permissions, you can still run it with, e.g. `bash my-script.sh`, but `. my-script.sh` will fail.

Looking at the code, the path to the `script` `source` file is passed as [an arg](https://github.com/argoproj/argo-workflows/blob/5b3909b9567a5c6fc4fef81ac0785cfddbffb291/workflow/controller/operator.go#L3002). I'm not entirely sure if `command` is set elsewhere, so it may very well depend on your container image's default `cmd` -- I think that would add up.

**Comment by vatine** [comment]
Certainly, the images that I can recall failing are on private repos (and, thus, as detailed in #11159 , the workflow controller cannot read the command from the image when constructing the pod), I have tried both `bash` and `bash -c` as the configured command in the configmap, with identical results. I can check what the pod command setting is. What I do know is that with this executable permission, it works. In all cases where this has failed, it is also not the first step in a workflow.

File: workflow/executor/executor.go [hunk 2]
```diff
@@ -267,11 +267,13 @@ func (we *WorkflowExecutor) LoadArtifacts(ctx context.Context) error {
 func (we *WorkflowExecutor) StageFiles() error {
 	var filePath string
 	var body []byte
+	mode := 0o644
```
Comment (agilgur5) [inline workflow/executor/executor.go:270]: this is failing to build due to a typecheck error (see [lint](https://github.com/argoproj/argo-workflows/actions/runs/8254003171/job/22577149813?pr=12787) failure etc).

I think this is supposed to be written as:
```suggestion
	mode := os.FileMode(0o644)
```

I was hoping there might be some constants for these modes, but seems like [no](https://go.dev/src/io/fs/fs.go) (note that [`os.FileMode = fs.FileMode`](https://pkg.go.dev/os#FileMode))

File: workflow/executor/executor.go [hunk 3]
```diff
@@ -267,11 +267,13 @@ func (we *WorkflowExecutor) LoadArtifacts(ctx context.Context) error {
...
 	switch we.Template.GetType() {
 	case wfv1.TemplateTypeScript:
 		log.Infof("Loading script source to %s", common.ExecutorScriptSourcePath)
 		filePath = common.ExecutorScriptSourcePath
 		body = []byte(we.Template.Script.Source)
+		mode = 0o755
```
Comment (agilgur5) [inline workflow/executor/executor.go:276]:
```suggestion
		mode = os.FileMode(0o755)
```

same as above

**Comment by vatine** [comment]
This is the command and args from the relevant pod:

```
$ kubectl get pod -n argo <pod name> -o json | jq '.spec.containers[1] | (.command,.args)'
[
  "/var/run/argo/argoexec",
  "emissary",
  "--loglevel",
  "debug",
  "--log-format",
  "text",
  "--"
]
[
  "/argo/staging/script"
]
```

I still cannot explain why it works in some cases and fail in others.

**Comment by agilgur5** [comment]
> I still cannot explain why it works in some cases and fail in others.

Sorry for the delayed response; I wanted to take some time to see if we could root cause this and have been pretty overloaded. Thanks for looking more into this.

> I have tried both `bash` and `bash -c` as the configured command in the configmap, with identical results.

This is surprising... based on this and your main container's `command`, it almost seems like it's not picking that `command` up from your configmap? If it works in some but not others, I might guess some images are not specified?

> In all cases where this has failed, it is also not the first step in a workflow.

This shouldn't matter, it has no impact on the `command`.

> What I do know is that with this executable permission, it works.

Since we don't know the mode of the `command` beforehand, I think this might still be a correct change; some `command`s will need this permission.
Since it's a permission though, I want to be pretty careful to not extend it in a way that could cause a vuln. So far, most threat models I've thought of are not particularly impactful since the `script` is user input anyway, but I'm not sure I've covered all the bases.

**Comment by vatine** [comment]
This is what Argo puts in the pod's "args:" and "command:" (note this probably has some relevance to issue 11159, as I see no actual trace of the "command" or "entrypoint" in the argoexec call; for those where t just seems to work,  can see a terminal "/bin/sh" on the argoexec command line).

``` 
$ kubectl get pods -n argo pr-run-e-e-tes-202404191144-docker-1590034149 -o json | jq '.spec.containers[1] | .command, .args'
[
  "/var/run/argo/argoexec",
  "emissary",
  "--loglevel",
  "info",
  "--log-format",
  "text",
  "--"
]
[
  "/argo/staging/script"
]
```

This is with the configmap configured to have that image mapped to:
```
cmd:
  - /bin/bash
```

**Comment by agilgur5** [comment]
> Based on my analysis in [#11159 (comment)](https://github.com/argoproj/argo-workflows/issues/11159#issuecomment-2066761200), since we don't know the mode of the `command` beforehand, I think this might still be a correct change; some `command`s will need this permission.

Apparently, per #13549, there was a separate issue filed exclusively for the script exec perms in #3496. Given both of those and the above, this is probably worthwhile to have in its own right, regardless of the Emissary nuances

**Review by agilgur5 — APPROVED** [review]
Thanks for finding, fixing, and iterating on this again!

**Comment by agilgur5** [comment]
> > I have tried both `bash` and `bash -c` as the configured command in the configmap, with identical results.
> 
> This is surprising... based on this and your main container's `command`, it almost seems like it's not picking that `command` up from your configmap?

> This is with the configmap configured to have that image mapped to:
> 
> ```yaml
> cmd:
>   - /bin/bash
> ```

I found the problem here. `cmd` in Docker is [equivalent to `args`](https://github.com/argoproj/argo-workflows/blob/54621cc60117cf68183be24322119d85a80bb650/workflow/controller/workflowpod.go#L403) in k8s. There's no example of this in the docs, but you could [set `entrypoint` instead](https://github.com/argoproj/argo-workflows/blob/a0ba3c70d18227e74bed1d77aa86c4ad4212d159/workflow/controller/entrypoint/config_index.go#L16) in your `images` config.

The bug is that for `script` templates, `cmd` seems to be accidentally ignored. The `entrypoint` + `cmd` lookup only [sets `args` if they are not already set](https://github.com/argoproj/argo-workflows/blob/54621cc60117cf68183be24322119d85a80bb650/workflow/controller/workflowpod.go#L402) in the `container` spec. And for `script` templates, the `ScriptSourcePath`, `/argo/staging/script` currently, is [appended to the `args`](https://github.com/argoproj/argo-workflows/blob/6b16f24c97e4b9c792e18835304e517f2d0ddd0f/workflow/controller/operator.go#L3023) just before Pod creation -> entrypoint lookup occurs.

This is fixable by changing the ordering, but may be a breaking change for `script` templates, as existing ones would unexpectedly have their `cmd` added.

For clarity, the full `command` + `args` for `script` templates would end up as:
```yaml
[
  "/var/run/argo/argoexec",
  "emissary",
  "--loglevel",
  "info",
  "--log-format",
  "text",
  "--",
  "<entrypoint>"
],
[
  "<cmd>",
  "/argo/staging/script"
]
```

where `cmd` is currently ignored.

**Pull request merged**